website/www/site/content/en/documentation/ml/anomaly-detection.md

```diff
@@ -0,0 +1,230 @@
+
+To have a better understanding of the dataset, here are some examples from the train split of the dataset:
+
+
+| Text        | Type of emotion |
+| :---        |    :----:   |
+| im grabbing a minute to post i feel greedy wrong      | Anger       |
+| i am ever feeling nostalgic about the fireplace i will know that it is still on the property   | Love        |
+| ive been taking or milligrams or times recommended amount and ive fallen asleep a lot faster but i also feel like so funny | Fear |
+| on a boat trip to denmark | Joy |
+| i feel you know basically like a fake in the realm of science fiction | Sadness |
+| i began having them several times a week feeling tortured by the hallucinations moving people and figures sounds and vibrations | Fear |
+
+### Anomaly Detection Algorithm
+[HDBSCAN](https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html) is a clustering algorithm which extends DBSCAN by converting it into a hierarchical clustering algorithm, and then using a technique to extract a flat clustering based in the stability of clusters. Once trained, when predicting the cluster for a new data point if outlier will output -1 else will predict one of the existing clusters.
+
+
+## Ingestion to PubSub
+We first ingest the data into [PubSub](https://cloud.google.com/pubsub/docs/overview) so that while clustering we can read the tweets from PubSub. PubSub is a messaging service for exchanging event data among applications and services. It is used for streaming analytics and data integration pipelines to ingest and distribute data.
+
+The full example code for ingesting data to PubSub can be found [here](sdks/python/apache_beam/examples/inference/anomaly_detection/write_data_to_pubsub_pipeline/)
+
+The file structure for ingestion pipeline is:
+
+    write_data_to_pubsub_pipeline/
+    ├── pipeline/
+    │   ├── __init__.py
+    │   ├── options.py
+    │   └── utils.py
+    ├── __init__.py
+    ├── config.py
+    ├── main.py
+    └── setup.py
+
+`pipeline/utils.py` contains the code for loading the emotion dataset and two `beam.DoFn` that are used for data transformation
+`pipeline/options.py` contains the pipeline options to configure the Dataflow pipeline
+`config.py` defines some variables like GCP PROJECT_ID, NUM_WORKERS that are used multiple times
+`setup.py` defines the packages/requirements for the pipeline to run
+`main.py` contains the pipeline code and some additional function used for running the pipeline
+
+### How to Run the Pipeline ?
+First, make sure you have installed the required packages.
+
+1. Locally on your machine: `python main.py`
+2. On GCP for Dataflow: `python main.py --mode cloud`
```

Comment:
We should probably include a step here for updating any variables in config.py (e.g. gcp project)

Reply:
(same comment applies in the other How to Run section)

Reply:
The variables that are in `config.py` are something that one won't change frequently in my opinion. Those are the variables that are usually configured once at the start and then usually not updated. So, I am not sure if it really helps.

Reply:
Okay adding that in documentation